v-next/hardhat-node-test-reporter/src/error-formatting.ts

```diff
@@ -37,44 +46,99 @@ export function formatError(error: Error): string {
 
   error = cleanupTestFailError(error);
 
-  const defaultFormat = inspect(error);
-  const indexOfMessage = defaultFormat.indexOf(error.message);
+  return formatSingleError(error);
+}
 
-  let title: string;
-  let stack: string;
-  if (indexOfMessage !== -1) {
-    title = defaultFormat.substring(0, indexOfMessage + error.message.length);
-    stack = defaultFormat
-      .substring(indexOfMessage + error.message.length)
-      .replace(/^(\r?\n)*/, "");
-  } else {
-    title = error.message;
-    stack = error.stack ?? "";
+/**
+ * This function takes an error and formats it into a human-readable string.
+ *
+ * The error is formatted as follows:
+ * - The error message is the first line of the error stack if the stack
+ *   contains the first line of the original error message, otherwise it's the
+ *   the first line of the original error message. The error message is prefixed
+ *   with the prefix, if provided. The error message is printed in red if it's
+ *   the first error in the error chain, otherwise it's printed in grey.
+ * - If the error is diffable (i.e. it has `actual` and `expected` properties),
+ *   the diff is printed.
+ * - The error stack is formatted as a series of references (lines starting with
+ *   "at"). The location part of the stack is normalized. The stack is printed
+ *   in grey, indented by 4 spaces.
+ * - If the error has a cause, the cause is formatted in the same way,
+ *   recursively. Formatting a cause increases the depth by 1. If the depth is
+ *   greater than 3, the cause is replaced by an error indicating that the error
+ *   chain has been truncated. The formatted cause is printed in grey, indented
+ *   by 2 spaces.
+ * - If the error is an aggregate, all the errors in the aggregate are formatted
+ *   in the same way, recursively. Then, they are printed in grey, indented by 2
+ *   spaces.
+ *
+ * @param error - The error to format
+ * @param prefix - A prefix to add to the error message
+ * @param depth - The depth of the error in the error chain
+ * @returns The formatted error
+ */
+function formatSingleError(
+  error: Error,
+  prefix: string = "",
+  depth: number = 0,
+): string {
+  const stackLines = (error.stack ?? "").split("\n");
+
+  let message = error.message.split("\n")[0];
+  if (stackLines.length > 0 && stackLines[0].includes(message)) {
+    message = stackLines[0];
   }
+  message = message.replace(" [ERR_ASSERTION]", "").replace(/:$/, "");
 
-  title = improveNodeAssertTitle(title, error);
-  title = chalk.red(title);
-  stack = replaceFileUrlsWithRelativePaths(stack);
-  stack = chalk.gray(stack);
+  if (prefix !== "") {
+    message = `[${prefix}]: ${message}`;
+  }
 
-  const diffResult = getErrorDiff(error);
+  const diff = getErrorDiff(error);
+
+  const stackReferences: StackReference[] = stackLines
+    .map(parseStackLine)
+    .filter((reference) => reference !== undefined);
```

Comment:
Should we really filter these out, print them "as is", or in a slightly different format?